dev/src/bulk-writer.ts

```diff
@@ -211,25 +224,97 @@ class BulkCommitBatch {
     // Capture the error stack to preserve stack tracing across async calls.
     const stack = Error().stack!;
 
-    return this.writeBatch.bulkCommit().catch(err => {
-      throw wrapError(err, stack);
-    });
+    let originalIndexMap: Map<number, number> = new Map(
+      Array.from(new Array(this.opCount), (_, i) => [i, i])
+    );
+
+    let retryAttempts = 0;
+    while (retryAttempts < MAX_RETRY_ATTEMPTS) {
+      let retryIndexes: number[] = [];
+      try {
+        await this.backoff.backoffAndWait();
+        const results = await this.writeBatch.bulkCommit();
+        retryIndexes = this.processResults(originalIndexMap, results);
```

Comment:
Can we rewrite this so that `processResults` doesn't take `originalIndexMap`? It seems like we could apply the  `retryIndexes` indexes in this function. I haven't played with this, but I wonder if this would lead to a simplification.

Reply:
It's possible to not pass in `originalIndexMap` to `processResults` -- the successful results are located at the indexes not in `retryIndexes`. From there we could resolve/reject the individual promises and reject the rest.

However, the main issue is that the `resultsMaps` resolution logic would have to live in `bulkCommit()`, which then means that `processResults()` would become `findRetryableIndexes()`. From there, it would make more sense to have all the logic in `bulkCommit`, which would make the method much longer. 

With the new refactor, using `originalIndexMap` in `processResults` seems a lot cleaner. WDYT?